keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,924 @@
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// 
+	// +optional
+	SucceededIndexes []int
+...
+}
+```
+
+However, if we allow users to set all `succeededIndexes` (`0-10^5`) to `.spec.successPolicy.creteria.succeededIndexes`
+and don't limit the number of list sizes, there are cases that the size of a job object is too big.
+
+In the worst case, allowed all indexes (`0-10^5`) are added to `succeededIndexes`, and a `succeededIndexes` size
+is `SUM[9*10^n*(n+1)]+2+6≈5.6656MiB`, where:
+- `n` starts from `0` and goes up to `5`.
+- `1` of `n+1` means a separator that indicates `,`.
+- `2` is the sum of the index `0` and `,`.
+- `6` is the size of indexes `10^5`.
+
+So, if we select this alternative API design, we need to limit the size of `succeededIndexes`. 
+
+### Acceptable percentage of total succeeded indexes in the succeededCount field
+
+We can accept a percentage of total succeeded indexes in the `succeededCount` field for the job with autoscaling semantics.
+However, there is no effective use case for typical stories using elastic horovod or PyTorch elastic training,
+as all pods must be completed.
+
+```golang
+// SuccessPolicyCriterion describes a Job can be succeeded based on succeeded indexes.
+type SuccessPolicyCriterion struct {
+	...
+	// Specifies the required number of indexes when .spec.completionMode =
+	// "Indexed".
+	// Value can be an absolute number (ex: 5) or an absolute percentage of total indexes
+	// when the job can be declared as succeeded (ex: 50%).
+	// The absolute number is calculated from the percentage by rounding up.
+	// 
+	// +optional
+	SucceededCount *intstr.IntOrString
+	...
+}
+```
+
+### Match succeededIndexes using CEL
+
+We can accept a set of required indexes represented by CEL in the `succeededIndexes` field. 
+However, it is difficult to represent the set of indexes without regularity. 
+
+```golang
+// SuccessPolicyCriterion describes a Job can be succeeded based on succeeded indexes.
+type SuccessPolicyCriterion struct {
+	...
+	// Specifies a set of required indexes using CEL.
+	// For example, if the completed indexes are only the last index, they are
+	// represented as (job.completions -1).
+	//
+	// +optional
+	SucceededIndexesMatchExpression *string
+	...
+}
+```
+
+### Use JobSet instead of Indexes Job
+
+The [JobSet](https://github.com/kubernetes-sigs/jobset) is a custom resource for managing a group of Job as a unit.
+
+Some of the stories are better served using JobSet.
+Specifically, cases that make assumptions about what an index represents could be mapped as jobs in JobSet,
+with names representing the semantics of those different groups of pods.
+
+However, both Job level and JobSet level successPolicies would be valuable
+since there are some cases in which we want to launch a Job by a single podTemplate.
+
+<<[UNRESOLVED @tenzen-y: Do we need to make configurable actions? ]>>
```

Comment:
I see a lot of unresolved sections here. I would probably remove these since these were alternatives considered. It looks a bit incomplete when reading the KEP from top-to-bottom.

Reply:
this is a non blocking nit but its something that bothered me a little.

Reply:
Actually, we need to discuss these sections for the second alpha stage in the future.
If we decide that we don't have the second alpha stage and don't include these features in this enhancement, I think we can remove `<<UNRESOLVED>>`.

Reply:
For the purpose of this iteration, these are not unresolved questions. We have concluded that they might be re-evaluated in a future version. You can mention this in the alpha/beta criteria, but you shouldn't leave the UNRESOLVED tags.

Reply:
I see. My understanding of the `<<UNRESOLVED>>` tag isn't correct.
Let me move `Future work` section.